azavea/pfb-network-connectivity
# Feature/kak/pass city fips#608

## Overview

Add the city FIPS to the neighborhood model, and pass and use it through job processing and in UI.


### Demo

The web UI neighborhood list page, now with added FIPS codes:
![pfb_neighborhood_list_with_fips](https://user-images.githubusercontent.com/960264/52089227-17fade00-257c-11e9-9479-0f1a4a4f54fc.png)

The web UI neighborhood creation page, with front-end validation for the FIPS:
![pfb_city_fips_form_error](https://user-images.githubusercontent.com/960264/52089267-2fd26200-257c-11e9-86d1-ce77a7e411bc.png)

Some custom errors the API will return on checking the FIPS:
![pfb_city_fips_start_state_error](https://user-images.githubusercontent.com/960264/52089293-3d87e780-257c-11e9-902d-eab01d95f082.png)

![pfb_city_fips_error_numeric](https://user-images.githubusercontent.com/960264/52089305-41b40500-257c-11e9-9c1c-c7cb14482711.png)

Development log output after kicking off a batch job. FIPS included in the command:
![pfb_django_batch_logs](https://user-images.githubusercontent.com/960264/52089349-585a5c00-257c-11e9-98d4-f856c0af60af.png)

During analysis run for one of the jobs, the default city speed was found (meaning FIPS passed correctly):
![pfb_boulder_found_fips_speed](https://user-images.githubusercontent.com/960264/52089392-71fba380-257c-11e9-9912-1203597d66b4.png)


### Notes

I wrote a script to look up city FIPS from [this Census list of FIPS for place names](https://www2.census.gov/geo/docs/reference/codes/files/national_places.txt) and add them to a given batch analysis shapefile, and ran it on the 100 cities batch file from S3. [The result is on S3 now](https://s3.amazonaws.com/pfb-batch-run-inputs-us-east-1/PFB_100_cities_cityfips.zip). I also made a test file with three of the 100 cities (Boulder, Sioux Falls, and New Haven, picked because they're small-ish.) [Test batch shapefile here](https://s3.amazonaws.com/pfb-batch-run-inputs-us-east-1/PFB_test_cities_cityfips.zip).

We should probably run the script on the other batch shapefiles in that bucket, if the one I've made already looks good.


## Testing Instructions

 * run migrations
 * should be able to create neighborhood in web UI or DRF UI, with or without city FIPS
 * should get appropriate error message for a bad FIPS
 * running analysis individually or in batch should use and pass the city FIPS, if present
 * logged analysis command should include FIPS, if present on the neighborhood
 * city speed should be found by city FIPS, if defined in `city_fips_speed.csv`


Closes #608
